### PR TITLE
feat: Searcher, Matcher and Dealer

### PR DESCRIPTION
**Pull request by sshaman1101**
Added:
- Searcher interface that hides Market/DWH communication (now for BIDs only).
- Matcher interface that picks up the best order. Split implementations for bid and ask.
- Dealer interface that handles dealing based on two orders. Now can perform full deal cycle from Node's side.

### CODE REVIEW

File: insonmnia/dealer/match.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+	return &bidMatcher{}
+}
+
+func (m *bidMatcher) Match(orders []*sonm.Order) (*sonm.Order, error) {
+	if len(orders) == 0 {
+		return nil, nil
+	}
+
+	sort.Sort(byPrice(orders))
+	return orders[0], nil
```
Comment (noxiouz) [inline insonmnia/dealer/match.go]: Oh! Use max. Seems u worked at HP

File: insonmnia/dealer/match.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+	return &askMatcher{}
+}
+
+func (m *askMatcher) Match(orders []*sonm.Order) (*sonm.Order, error) {
+	if len(orders) == 0 {
+		return nil, nil
+	}
+
+	sort.Sort(byPrice(orders))
+	return orders[len(orders)-1], nil
```
Comment (noxiouz) [inline insonmnia/dealer/match.go]: min :)

File: insonmnia/dealer/searcher.go [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+// SearchFilter represent params for searching and filtering orders
+// on the marketplace. Accepted by `Searcher` interface.
+type SearchFilter struct {
+	order     *sonm.Order
+	balance   *big.Int
+	allowance *big.Int
+}
+
+// NewSearchFilter validates input data and constructs `SearchFilter`
+func NewSearchFilter(order *sonm.Order, bal, alw *big.Int) (*SearchFilter, error) {
```
Comment (3Hren) [inline insonmnia/dealer/searcher.go]: Come on. Use normal variable names.

**Comment by 3Hren** [comment]
Closed in favor of #515.

**Pull request closed without merging**